xcm/xcm-builder/src/tests.rs

```diff
@@ -0,0 +1,338 @@
+		&ConcreteFungible { id: X1(Parent), amount: 100 },
+		&X1(Parent),
+	));
+
+	assert_eq!(to_account(X1(Parachain{id:1})), Ok(1001));
+	assert_eq!(to_account(X1(Parachain{id:50})), Ok(1050));
+	assert_eq!(to_account(X2(Parent, Parachain{id:1})), Ok(2001));
+	assert_eq!(to_account(X2(Parent, Parachain{id:50})), Ok(2050));
+	assert_eq!(to_account(X1(AccountIndex64{index:1, network:Any})), Ok(1));
+	assert_eq!(to_account(X1(AccountIndex64{index:42, network:Any})), Ok(42));
+	assert_eq!(to_account(Null), Ok(3000));
+}
+
+#[test]
+fn weigher_should_work() {
+	let mut message = opaque::Xcm::ReserveAssetDeposit {
+		assets: vec![ConcreteFungible { id: X1(Parent), amount: 100 }],
+		effects: vec![
+			Order::BuyExecution { fees: All, weight: 0, debt: 30, halt_on_error: true, xcm: vec![] },
+			Order::DepositAsset { assets: vec![All], dest: Null },
+		],
+	}.into();
+	assert_eq!(<TestConfig as Config>::Weigher::shallow(&mut message), Ok(30));
+}
+
+#[test]
+fn take_weight_credit_barrier_should_work() {
+	let mut message = opaque::Xcm::TransferAsset {
+		assets: vec![ConcreteFungible { id: X1(Parent), amount: 100 }],
+		dest: Null,
+	};
+
+	let mut weight_credit = 10;
+	let r = TakeWeightCredit::should_execute(
+		&X1(Parent),
+		true,
+		&mut message,
+		10,
+		&mut weight_credit,
+	);
+	assert_eq!(r, Ok(()));
+	assert_eq!(weight_credit, 0);
+
+	let r = TakeWeightCredit::should_execute(
+		&X1(Parent),
+		true,
+		&mut message,
+		10,
+		&mut weight_credit,
+	);
+	assert_eq!(r, Err(()));
+	assert_eq!(weight_credit, 0);
+}
+
+#[test]
+fn allow_unpaid_should_work() {
+	let mut message = opaque::Xcm::TransferAsset {
+		assets: vec![ConcreteFungible { id: X1(Parent), amount: 100 }],
+		dest: Null,
+	};
+
+	AllowUnpaidFrom::set(vec![ X1(Parent) ]);
+
+	let r = AllowUnpaidExecutionFrom::<IsInVec<AllowUnpaidFrom>>::should_execute(
+		&X1(Parachain { id: 1 }),
+		true,
+		&mut message,
+		10,
+		&mut 0,
+	);
+	assert_eq!(r, Err(()));
+
+	let r = AllowUnpaidExecutionFrom::<IsInVec<AllowUnpaidFrom>>::should_execute(
+		&X1(Parent),
+		true,
+		&mut message,
+		10,
+		&mut 0,
+	);
+	assert_eq!(r, Ok(()));
+}
+
+#[test]
+fn allow_paid_should_work() {
+	AllowPaidFrom::set(vec![ X1(Parent) ]);
+
+	let mut message = opaque::Xcm::TransferAsset {
+		assets: vec![ConcreteFungible { id: X1(Parent), amount: 100 }],
+		dest: Null,
+	};
+
+	let r = AllowTopLevelPaidExecutionFrom::<IsInVec<AllowPaidFrom>>::should_execute(
+		&X1(Parachain { id: 1 }),
+		true,
+		&mut message,
+		10,
+		&mut 0,
+	);
+	assert_eq!(r, Err(()));
+
+	let mut underpaying_message = opaque::Xcm::ReserveAssetDeposit {
+		assets: vec![ConcreteFungible { id: X1(Parent), amount: 100 }],
+		effects: vec![
+			Order::BuyExecution { fees: All, weight: 0, debt: 20, halt_on_error: true, xcm: vec![] },
+			Order::DepositAsset { assets: vec![All], dest: Null },
+		],
+	};
+
+	let r = AllowTopLevelPaidExecutionFrom::<IsInVec<AllowPaidFrom>>::should_execute(
+		&X1(Parachain { id: 1 }),
```

Comment:
I believe the intention here is to show that with the correct origin (Parent), you cannot underpay.

```suggestion
		&X1(Parent),
```